### ISSUE TITLE
Reading side of async logger busy waits if log messages are pushed frequently

### ISSUE THREAD

**Issue by Weeena** [gabime/spdlog#207]
When asynchronsous logging is used and If log messages are pushed frequently, the reading side of the asynchronous logger is busy waiting. The reason for this is the heuristic in async_log_helper::sleep_or_yield(), which polls without any pause if the last message was read < 1ms before and which does a yield() if the last message was read < 10ms before. (Note that this yield will only then lead to some other thread being scheduled if there are any threads with the same priority.)

In our application, logs occur very frequently (several log messages each ms) which leads to the asynchronous logger's reading thread claiming 100% core time.

I don't have a solution for this, but to me it seems that the heuristic needs to fit to the application. Maybe it should be possible to control the used heuristic? Of course, as an alternative it would be nice if the reading side of the asynchronous logger would be woken up if a new log entry is put in an empty queue or if the number of log entries exceed some threshold.

I would be glad to hear from you and am curious about your thoughts on this topic.

**Comment by gabime** [comment]
I agree with you. it is application dependant. especially with high rate logging apps like yours.
The simplest in your case seem to reduce the 1ms in the spdlog code (line #337) to few hundered micros instead, and see if it helps.. another option would be to remove the spin line in the code completely.
please try those options and report what best works. after that we can see how to make those numbers configurable (maybe add to spdlog's tweakme.h file params to decide to spin/yield/sleep intervals.. not sure yet)

**Comment by gabime** [comment]
Fixed (partially) in ea611f2d792b3c492451c8965eda183501cfb175

**Comment by Weeena** [comment]
Thank you, the changed heuristic for the asynchronous logger's reader (your commit "reduced spinning duation in async_log_helper") works better for our application than the former one. Nevertheless, it is still a heuristic and maybe will not fit to every application. Additionally, the `yield()` will only then lead to an actual yield if there are other worker threads of the same priority as the logger. In our environment, the high frequency of the `yield()` makes the scheduler rather busy which influences the overall system performance.

In our opinion, it would be ideal if the asynchronous logger's reader would not be busy waiting but rather awakened if there is work to do. Maybe it could be awakened if the queue's fill level exceeds some threshold and, additionally, if some timeout is exceeded.

Maybe the dilemma could also be solved with the use of a different queue. I recently came across a highly recommended lockless queue (see [concurrentqueue](https://github.com/cameron314/concurrentqueue)). Maybe this queue fits to your library (according to its author, its fast and header-only). Maybe the blocking version (where the consumer side can wait with a timeout) would be an appropriate choice. Unfortunately, I haven't used that queue myself so everything I write here is just from information I read.

What is your opinion? Shall I open a new issuewith the text of this comment? Or would you like to reopen the old issue? Or do you have other plans?

**Comment by gabime** [comment]
I tried once this queue. It totally not suitable for logging since the messages get popped in a weird order and end up in a mess in the final log.

About  the yield/spin, please note it last for total of just 100 MICROS. Is it really that of a problem? what platform are you using ?

I do have a plan to add a conventional mutex to wake up waiting consumer/producers but i want it to play nicely with the lockless feature.. (i.e. still preserve the lockless feature and use the locking after some spinning.. probably..)

**Comment by Weeena** [comment]
Oh, I see - with that queue obviously speed has a too high price ;-( ...

What we observed was the following: In a test setup, we had some worker tasks that logged some messages. With the old heuristic of the asynchronous logger's reader we observed that the processing time for the worker tasks was rather distributed (if I remember correctly between 10 to 20 ms). We changed the heuristic such that the asynchronous logger's reader did neither poll nor yield, just sleep for 20 ms after doing its work (which was a heuristic sleep time that fit to our log frequency and queue size). After that change, the worker task's processing times were much more stable.
Our Environment: Linux, ARMv8 (= AArch64), gcc 4.9.
